### PR TITLE
Avoid deref of Ref in copy constructor of PackedCandidate

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Previously if the vertex had not been read from the original, a call to a copy constructor would cause the deref to happen.

The copy constructor was being called by ROOT during read back from a file. This meant while we were reading the branch holding the PackedCandidate we were trigger reading the branch holding the Vertexes.

#### PR validation:

The code compiles and the extended unit test passes.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Please test

**Comment by slava77** [comment]
> Previously if the vertex had not been read from the original, a call to a copy constructor would cause the deref to happen.
> 
> The copy constructor was being called by ROOT during read back from a file. This meant while we were reading the branch holding the PackedCandidate we were trigger reading the branch holding the Vertexes.

is there a chance that  a `    PackedCandidate(PackedCandidate &&iOther)` is called or any of the assignment operators?

File: DataFormats/PatCandidates/interface/PackedCandidate.h [hunk 1]
```diff
@@ -204,12 +204,12 @@ namespace pat {
...
-          dz_(iOther.dz_),
-          dphi_(iOther.dphi_),
-          deta_(iOther.deta_),
-          dtrkpt_(iOther.dtrkpt_),
+          vertex_((iOther.vertex_ ? new Point(iOther.vertex()) : nullptr)),
+          dxy_(vertex_ ? iOther.dxy_ : 0),
+          dz_(vertex_ ? iOther.dz_ : 0),
+          dphi_(vertex_ ? iOther.dphi_ : 0),
+          deta_(vertex_ ? iOther.deta_ : 0),
+          dtrkpt_(vertex_ ? iOther.dtrkpt_ : 0),
```
Comment (slava77) [inline DataFormats/PatCandidates/interface/PackedCandidate.h:212]: this doesn't look right; having a null vertex does not imply zero values in these parameters
https://github.com/cms-sw/cmssw/blob/22f5521804d2e9a1ffee48aeefecca1bef226295/DataFormats/PatCandidates/src/PackedCandidate.cc#L122-L132
I think that the unpacking of parameters not dependent on the value of the `vertexRef` should still happen.

Comment (slava77) [reply DataFormats/PatCandidates/interface/PackedCandidate.h:212]: uhm, on a second thought, perhaps not and the changes are OK, but that becomes a bit more obvious only after seeing that all cases to access the `dxy_` through `dtrkpt_` values rely on having `unpackVtx` called.

I see a path after this PR change that a candidate that comes out of this `PackedCandidate(const PackedCandidate &iOther)`  with `iOther.vertex_ == nullptr` can get to e.g. a call to `::packVtx` without a preceding `unpackVtx` (e.g. via `::setTrackProperties`) and lead to a crash in 
https://github.com/cms-sw/cmssw/blob/22f5521804d2e9a1ffee48aeefecca1bef226295/DataFormats/PatCandidates/src/PackedCandidate.cc#L26-L29

This looks like a `maybeUnpackBoth()` call is needed in `::setTrackProperties`

**Comment by Dr15Jones** [comment]
> is there a chance that a ` PackedCandidate(PackedCandidate &&iOther)` is called or any of the assignment operators?

Those methods either already avoid reading the `vertex_` or only do so if `iOther.vertex_` is not null. Now they do read such values as `iOther.dphi_` which are a race condition if `iOther.vertex_ == nullptr`. However, those values (which are not guaranteed to be reasonable in that case) will never be used as the object will have `vertex_ == nullptr` which will trigger unpacking.

File: DataFormats/PatCandidates/interface/PackedCandidate.h (hunk 1 above)
Comment (Dr15Jones) [reply DataFormats/PatCandidates/interface/PackedCandidate.h:212]: Good catch! I made the change.

**Comment by Dr15Jones** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32964	85dfd14
- code changes are in line with the PR description and the follow up review; the physics outputs should not change
- jenkins tests pass and comparisons with the baseline show no relevant differences as expected
    - a few workflows like 140.56 have `ExcessiveTime` LogWarning in the PR tests for the same module in all 4 cases that I checked, with the message looking like 
```
%MSG-e ExcessiveTime:  SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  24-Feb-2021 16:51:38 CET Run: 326479 Event: 1394020
ExcessiveTime: Module used 614.385 seconds of time which exceeds the error threshold configured in the Timing Service of 600 seconds.

```
As I recall in the past similar symptoms were an indication of some issues in the Frontier infrastructure

**Comment by silviodonato** [comment]
+1

**Pull request merged**